api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java

```diff
@@ -1,16 +1,120 @@
+import org.pmiops.workbench.mandrill.model.MandrillMessage;
+import org.pmiops.workbench.mandrill.model.MandrillMessageStatuses;
+import org.pmiops.workbench.mandrill.model.MandrillMessageStatus;
+import org.pmiops.workbench.mandrill.model.RecipientAddress;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
+import java.util.Collections;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.inject.Provider;
 import javax.mail.Message;
 import javax.mail.MessagingException;
 import javax.mail.Transport;
 
 @Service
 public class MailServiceImpl implements MailService {
 
+    private final Provider<MandrillApi> mandrillApiProvider;
+    private final Provider<CloudStorageService> cloudStorageServiceProvider;
+    private Provider<WorkbenchConfig> workbenchConfigProvider;
+    private static final Logger log = Logger.getLogger(MailServiceImpl.class.getName());
+    enum Status {REJECTED, API_ERROR, SUCCESSFUL}
+
+    @Autowired
+    public MailServiceImpl(Provider<MandrillApi> mandrillApiProvider,
+                           Provider<CloudStorageService> cloudStorageServiceProvider,
+                           Provider<WorkbenchConfig> workbenchConfigProvider) {
+        this.mandrillApiProvider = mandrillApiProvider;
+        this.cloudStorageServiceProvider = cloudStorageServiceProvider;
+        this.workbenchConfigProvider = workbenchConfigProvider;
+    }
+
     public void send(Message msg) throws MessagingException {
         Transport.send(msg);
     }
 
+    public void sendWelcomeEmail(String contactEmail, String password, User user) throws MessagingException {
+        String apiKey = cloudStorageServiceProvider.get().readMandrillApiKey();
+        int retries = workbenchConfigProvider.get().mandrill.sendRetries;
+        MandrillApiKeyAndMessage keyAndMessage = new MandrillApiKeyAndMessage();
+        keyAndMessage.setKey(apiKey);
+        MandrillMessage msg = buildWelcomeMessage(contactEmail, password, user);
+        keyAndMessage.setMessage(msg);
+        retry: do {
+            retries--;
+            log.log(Level.INFO, "retries: " + retries);
+            ImmutablePair attempt = trySend(keyAndMessage);
+            Status status = Status.valueOf(attempt.getLeft().toString());
+            switch (status) {
+                case API_ERROR:
+                    log.log(Level.WARNING, String.format(
+                      "ApiException: Welcome Email to '%s' for user '%s' not sent: %s", contactEmail, user.getName(), attempt.getRight().toString()));
+                    if (retries == 0) {
+                        log.log(Level.SEVERE, String.format(
+                          "ApiException: On Last Attempt! Welcome Email to '%s' for user '%s' not sent: %s", contactEmail, user.getName(), attempt.getRight().toString()));
+                        throw new MessagingException("Sending email failed: " + attempt.getRight().toString());
```

Comment:
Why do we throw an exception in this case and the REJECTED case but not in the default case?

FWIW: I don't think we should return a 500 when we get back REJECTED here... as the input is valid, the user just gave an email address that we couldn't send stuff to. Seems better to just have the create user response indicate if there was a problem sending the email but return a 200?

Reply:
If I read this right, in the API error case, we don't actually know what went wrong here. This is an error from the api client talking to the mandrill server for some reason. In this case, I do think we have to throw an exception. How that is returned to the user is abstracted to, in this case, the `DirectoryServiceImpl` class.